crates/aptos/src/move_tool/mod.rs

```diff
@@ -853,6 +860,159 @@ impl CliCommand<String> for BuildPublishPayload {
+            object_address
+        );
+        prompt_yes_with_override(&message, self.txn_options.prompt_options)?;
+
+        let payload = aptos_cached_packages::aptos_stdlib::object_code_deployment_publish(
+            bcs::to_bytes(&package.extract_metadata()?).expect("PackageMetadata has BCS"),
+            package.extract_code(),
+        );
+        let size = bcs::serialized_size(&payload)?;
+        println!("package size {} bytes", size);
+
+        if !self.override_size_check && size > MAX_PUBLISH_PACKAGE_SIZE {
+            return Err(CliError::UnexpectedError(format!(
+                "The package is larger than {} bytes ({} bytes)! To lower the size \
+                you may want to include less artifacts via `--included-artifacts`. \
+                You can also override this check with `--override-size-check",
+                MAX_PUBLISH_PACKAGE_SIZE, size
+            )));
+        }
+        self.txn_options
+            .submit_transaction(payload)
+            .await
+            .map(TransactionSummary::from)
+    }
+}
+
+#[derive(Parser)]
+pub struct UpgradeObjectPackage {
+    /// Address of the object containing the package
+    #[clap(long, value_parser = crate::common::types::load_account_arg)]
+    pub(crate) object_address: AccountAddress,
+    /// Whether to override the check for maximal size of published data
+    ///
+    /// This won't bypass on chain checks, so if you are not allowed to go over the size check, it
+    /// will still be blocked from publishing.
+    #[clap(long)]
+    pub(crate) override_size_check: bool,
+    #[clap(flatten)]
+    pub(crate) included_artifacts_args: IncludedArtifactsArgs,
+    #[clap(flatten)]
+    pub(crate) move_options: MovePackageDir,
+    #[clap(flatten)]
+    pub(crate) txn_options: TransactionOptions,
+}
+
+#[async_trait]
+impl CliCommand<TransactionSummary> for UpgradeObjectPackage {
+    fn command_name(&self) -> &'static str {
+        "UpgradeObjectPackage"
+    }
+
+    async fn execute(self) -> CliTypedResult<TransactionSummary> {
+        let options = self
+            .included_artifacts_args
+            .included_artifacts
+            .build_options(
+                self.move_options.dev,
+                self.move_options.skip_fetch_latest_git_deps,
+                self.move_options.named_addresses(),
+                self.move_options.bytecode_version,
+                self.move_options.compiler_version,
+                self.move_options.skip_attribute_checks,
+                self.move_options.check_test_code,
+            );
+        let package = BuiltPackage::build(self.move_options.get_package_path()?, options)?;
+        let compiled_metadata = package.extract_metadata()?;
+        if compiled_metadata.upgrade_policy == UpgradePolicy::immutable() {
+            return Err(CliError::CommandArgumentError(
+                "A package with upgrade policy `immutable` cannot be upgraded".to_owned(),
+            ));
+        }
```

Comment:
This actually is problematic.

It assumes that the offchain data supercedes the onchain data.

You'll have to check the `onchain` policy, and not the compiled one.  Otherwise, you can't change a package to immutable.

Reply:
You're right, will amend this.